### ISSUE TITLE
How to override gradient in a FuncGraph created by tf.function

### ISSUE THREAD

**Issue by lffan** [tensorflow/tensorflow#36870]
With tensorflow 1.x API, I can override gradients with a gradient_override_map during building a graph as follows

```    
@tf.RegisterGradient('CustZero')
def _custom_grad(op, grad):
    return tf.zeros_like(op.inputs[0])
    
with tf.compat.v1.Session() as sess:
    with sess.graph.as_default() as g:
        x = tf.convert_to_tensor([-1.0, 0.0, 1.0, 2.0])
        with g.gradient_override_map({'Relu': 'CustZero'}):
            y = tf.nn.relu(x)
    dy = tf.gradients(y, x)

# dy: [0., 0., 0., 0.]
```

As we are moving to TF2, I'd like to use native TF2 API to do the same task with `tf.function` and `tf.GradientTape`. As I have learnt, `tf.function` creates a FuncGraph for each different `input_signature`, and I can use `get_concrete_function` to create one ConcreteFunction with a specific `input_signature`. Then I can get the `FuncGraph` associated with the `ConcreteFunction`. This is what I have tried, which failed to override gradients.

```
# create a concrete function with the specified input signatures
@tf.function(input_signature=(tf.TensorSpec(shape=(None,), dtype=tf.float32),))
def my_relu(x):
    return tf.nn.relu(x)
my_relu_conc = my_relu.get_concrete_function()

x = tf.constant([-1.0, 0.0, 1.0, 2.0], dtype=tf.float32)
with tf.GradientTape() as tape:
    with my_relu_conc.graph.gradient_override_map({'Relu': 'CustZero'}):
        tape.watch(x)
        y = my_relu_conc(x)
dy = tape.gradient(y, x).numpy()

# dy: [0., 0., 1., 1.]
```

From my understanding, since the FuncGraph is already traced, the graph cannot be changed and I cannot override with custom gradients. 

My question is: are there any methods to override gradient using native TF2 APIs?

If not, could we add an additional parameter named `gradient_override_map` to `tf.function` or `get_concrete_function` such that we can override gradients in the process where a FuncGraph is created? 

```
@tf.function(
    input_signature=(tf.TensorSpec(shape=(None,), dtype=tf.float32),), 
    gradient_override_map={'Relu': 'CustZero'})
def my_relu(x):
    return tf.nn.relu(x)
my_relu_conc = my_relu.get_concrete_function()
```
or 

```
@tf.function()
def my_relu(x):
    return tf.nn.relu(x)
my_relu_conc = my_relu.get_concrete_function(
    input_signature=(tf.TensorSpec(shape=(None,), dtype=tf.float32),), 
    gradient_override_map={'Relu': 'CustZero'})
```

**Comment by saxenasaurabh** [comment]
Could you try using `tf.custom_gradient`

```
@tf.function(input_signature=(tf.TensorSpec(shape=(None,), dtype=tf.float32),))
def my_relu(x):
  @tf.custom_gradient
  def relu_with_zeros_grad(x):
    output = tf.nn.relu(x)
    def grad_fn(grad):
       return tf.zeros_like(output.op.inputs[0])
    return output, grad_fn
  return relu_with_zeros_grad(x)
```

**Comment by sevenHsu** [comment]
???
